unified-scheduler-logic/src/lib.rs

```diff
@@ -1,20 +1,1333 @@
+#[derive(Debug)]
+pub struct TaskInner {
     transaction: SanitizedTransaction,
     index: usize,
+    lock_attempts: Vec<LockAttempt>,
+    blocked_page_count: TokenCell<ShortCounter>,
 }
 
-impl Task {
-    pub fn create_task(transaction: SanitizedTransaction, index: usize) -> Self {
-        Task { transaction, index }
-    }
-
+impl TaskInner {
     pub fn task_index(&self) -> usize {
         self.index
     }
 
     pub fn transaction(&self) -> &SanitizedTransaction {
         &self.transaction
     }
+
+    fn lock_attempts(&self) -> &Vec<LockAttempt> {
+        &self.lock_attempts
+    }
+
+    fn blocked_page_count_mut<'t>(
+        &self,
+        token: &'t mut BlockedPageCountToken,
+    ) -> &'t mut ShortCounter {
+        self.blocked_page_count.borrow_mut(token)
+    }
+
+    fn set_blocked_page_count(&self, token: &mut BlockedPageCountToken, count: ShortCounter) {
+        *self.blocked_page_count_mut(token) = count;
+    }
+
+    #[must_use]
+    fn try_unblock(self: &Task, token: &mut BlockedPageCountToken) -> Option<Task> {
+        self.blocked_page_count_mut(token)
+            .decrement_self()
+            .is_zero()
+            .then(|| self.clone())
+    }
+}
+
+/// [`Task`]'s per-address attempt to use a [page](Page) with [certain kind of
+/// request](RequestedUsage).
+#[derive(Debug)]
+struct LockAttempt {
+    page: Page,
+    requested_usage: RequestedUsage,
+}
+const_assert_eq!(mem::size_of::<LockAttempt>(), 16);
+
+impl LockAttempt {
+    fn new(page: Page, requested_usage: RequestedUsage) -> Self {
+        Self {
+            page,
+            requested_usage,
+        }
+    }
+
+    fn page_mut<'t>(&self, page_token: &'t mut PageToken) -> &'t mut PageInner {
+        self.page.0.borrow_mut(page_token)
+    }
+}
+
+/// Status about how the [`Page`] is used currently. Unlike [`RequestedUsage`], it has additional
+/// variant of [`Unused`](`PageUsage::Unused`).
+#[derive(Copy, Clone, Debug, Default)]
+enum PageUsage {
+    #[default]
+    Unused,
+    Readonly(ShortCounter),
+    Writable,
+}
+const_assert_eq!(mem::size_of::<PageUsage>(), 8);
+
+impl PageUsage {
+    fn from_requested_usage(requested_usage: RequestedUsage) -> Self {
+        match requested_usage {
+            RequestedUsage::Readonly => PageUsage::Readonly(ShortCounter::one()),
+            RequestedUsage::Writable => PageUsage::Writable,
+        }
+    }
+}
+
+/// Status about how a task is requesting to use a particular [`Page`]. Unlike [`PageUsage`],
+/// it has only two unit variants.
+#[derive(Clone, Copy, Debug)]
+enum RequestedUsage {
+    Readonly,
+    Writable,
+}
+
+/// Internal scheduling data about a particular address.
+///
+/// Specifially, it holds the current [`PageUsage`] (or no usage with [`PageUsage::Unused`]) and
+/// which [`Task`]s are blocked to be executed after the current task is notified to be finished
+/// via [`::deschedule_task`](`SchedulingStateMachine::deschedule_task`)
+#[derive(Debug)]
+struct PageInner {
```

Comment:
is there additional data that will be stored on these pages? I think this naming confused me previously. Wonder if something like:
- `AccessList`
- or `UsageList`
is more descriptive?

Reply:
@apfitzge (once we settled on this, I'll do the once for all renames; so pinging).

Reply:
Even with ping I somehow didn't get a github alert - sorry!

out of that list I think I prefer `UsageQueue` the most, or maybe even `LockQueue` (slightly shorter). Not as short as `Page` or `Lock` but I think the fact we use it as  a queue is an important description to be in the name.

Reply:
thanks for the update. kk, let's go with `UsageQueue`. :)

Reply:
this is done: 725e314efa65357ba9e670f298eacd16d4888b4e and 50b761652d268d837eb1f8b288a57c9f4dbd107e